source/agora/common/TransactionManager.d

```diff
@@ -0,0 +1,283 @@
+
+    public Output newTxOut (PublicKey address, Amount value = 0)
+    {
+        Output txOut;
+        txOut.address = address;
+        txOut.value = value;
+        return txOut;
+    }
+
+    /*******************************************************************************
+
+        Checks whether the transaction is coinbase
+
+        Params:
+            tx = Type of `Transaction`
+
+        Return:
+            Return true if this transaction is a coinbase
+
+    *******************************************************************************/
+
+    public bool isCoinbaseTx (Transaction tx)
+    {
+        return
+            (
+                (tx.inputs.length == 1) &&
+                (tx.inputs[0].previous == TransactionManager.nullTxHash) &&
+                (tx.inputs[0].index == 0) &&
+                (tx.inputs[0].signature == TransactionManager.nullSignature)
+            );
+    }
+
+    /*******************************************************************************
+
+        Creates a new coinbase transaction
+
+        Params:
+            address = The public key that can redeem this output
+            value = The initial value
+
+        Return:
+            Return coinbase transation
+
+    *******************************************************************************/
+
+    public Transaction newCoinbaseTX (PublicKey address, Amount value = 0)
+    {
+        Transaction tx;
+        tx.inputs ~= this.newTxIn(TransactionManager.nullTxHash, 0);
+        tx.outputs ~= this.newTxOut(address, value);
+        return tx;
+    }
+
+    /*******************************************************************************
+
+        Find a output of transaction
+
+        Params:
+            hash = `Hash` of `Transaction`
+            index = Index of the `Output` in the `Transaction`
+
+        Return:
+            `Output` in the `Transaction`
+
+    *******************************************************************************/
+
+    public Output * findTxOutput (Hash hash, uint index)
+    {
+        if (hash in this.txs)
+        {
+            auto tx = this.txs[hash];
+            if (index < tx.outputs.length)
+            {
+                return &tx.outputs[index];
+            }
+            else
+            {
+                return null;
+            }
+        }
+        else
+        {
+            return null;
+        }
+    }
+
+    /*******************************************************************************
+
+        Get sum of `Input`
+
+        Params:
+            tx = `Transaction`
+
+        Return:
+            Sum of `Input` in the `Transaction`
+
+    *******************************************************************************/
+
+    public long getInputSum (Transaction tx)
+    {
+        if (isCoinbaseTx(tx))
+            return 0;
+
+        long sum = 0;
+        Output *txOut;
+
+        foreach (idx, ref txIn; tx.inputs)
```

Comment:
I think you should take a look at [`std.algorithm`](https://dlang.org/phobos/std_algorithm.html).
Many of those loop could be turned into calls to `each`, `map` & co.